File: test/selector.js

Object.extend = function(destination, source) {
  for (var property in source)
    destination[property] = source[property];
  return destination;
};

Object.keys = function(object) {
  var results = [];
  for (var property in object)
    results.push(property);
  return results;
};


function $A(iterable) {
  if (!iterable) return [];
  // Safari <2.0.4 crashes when accessing property of a node list with property accessor.
  // It nevertheless works fine with `in` operator, which is why we use it here
  if ('toArray' in iterable) return iterable.toArray();
  var length = iterable.length || 0, results = new Array(length);
  while (length--) results[length] = iterable[length];
  return results;
}


var Class = (function() {
  function create() {
    var parent = null, properties = $A(arguments);
    if (Object.isFunction(properties[0]))
      parent = properties.shift();
      
    function klass() {
      this.initialize.apply(this, arguments);
    }
    
    Object.extend(klass, Class.Methods);
    klass.superclass = parent;
    klass.subclasses = [];
    
    if (parent) {
      var subclass = function() {};
      subclass.prototype = parent.prototype;
      klass.prototype = new subclass;
      parent.subclasses.push(klass);
    }
    
    for (var i = 0; i < properties.length; i++)
      klass.addMethods(properties[i]);
      
    if (!klass.prototype.initialize)
      klass.prototype.initialize = Prototype.emptyFunction;
      
    klass.prototype.constructor = klass;
    return klass;
  }
  
  function addMethods(source) {
    var ancestor = this.superclass && this.superclass.prototype;
    var properties = Object.keys(source);
    
    if (!Object.keys({ toString: true }).length) {
      if (source.toString != Object.prototype.toString)
        properties.push("toString");
      if (source.valueOf != Object.prototype.valueOf)
        properties.push("valueOf");
    }
                                                                 
    for (var i = 0, length = properties.length; i < length; i++) {
      var property = properties[i], value = source[property];
      if (ancestor && Object.isFunction(value) &&
          value.argumentNames().first() == "$super") {
        var method = value;
        value = (function(m) {
          return function() { return ancestor[m].apply(this, arguments); };
        })(property).wrap(method);
 
        value.valueOf = method.valueOf.bind(method);
        value.toString = method.toString.bind(method);
      }
      this.prototype[property] = value;
    }
    
    return this;
  }
  
  return {
    create: create,
    Methods: {
      addMethods: addMethods
    }
  };
})();


Object.extend(Object, {
  isFunction: function(object) {
    return typeof object === "function";
  },
  
  isString: function(object) {
    function getClass(object) {
      return Object.prototype.toString.call(object)
       .match(/^\[object\s(.*)\]$/)[1]; 
    }
    
    return getClass(object) === "String";
  }
});


RegExp.escape = function(str) {
  return String(str).replace(/([.*+?^=!:${}()|[\]\/\\])/g, '\\$1');
};


String.interpret = function(value) {
  return value == null ? '' : String(value);
};

Object.extend(String.prototype, {
  strip: function() {
    return this.replace(/^\s+/, '').replace(/\s+$/, '');
  },
  
  startsWith: function(pattern) {
    return this.indexOf(pattern) === 0;
  },
  
  gsub: function(pattern, replacement) {
    function prepareReplacement(replacement) {
      if (Object.isFunction(replacement)) return replacement;
      var template = new Template(replacement);
      return function(match) { return template.evaluate(match) };
    }
      
    var result = '', source = this, match;
    replacement = prepareReplacement(replacement);
    
    if (Object.isString(pattern))
      pattern = RegExp.escape(pattern);
  
    if (!(pattern.length || pattern.source)) {
      replacement = replacement('');
      return replacement + source.split('').join(replacement) + replacement;
    }
  
    while (source.length > 0) {
      if (match = source.match(pattern)) {
        result += source.slice(0, match.index);
        result += String.interpret(replacement(match));
        source  = source.slice(match.index + match[0].length);
      } else {
        result += source, source = '';
      }
    }
    return result;
  }
});



var Template = Class.create({
  initialize: function(template, pattern) {
    this.template = template.toString();
    this.pattern = pattern || Template.Pattern;
  },
  
  evaluate: function(object) {
    if (Object.isFunction(object.toTemplateReplacements))
      object = object.toTemplateReplacements();

    return this.template.gsub(this.pattern, function(match) {
      if (object == null) return '';
      
      var before = match[1] || '';
      if (before == '\\') return match[2];
      
      var ctx = object, expr = match[3];
      var pattern = /^([^.[]+|\[((?:.*?[^\\])?)\])(\.|\[|$)/;
      match = pattern.exec(expr);
      if (match == null) return before;

      while (match != null) {
        var comp = match[1].startsWith('[') ? match[2].gsub('\\\\]', ']') : match[1];
        ctx = ctx[comp];
        if (null == ctx || '' == match[3]) break;
        expr = expr.substring('[' == match[3] ? match[1].length : match[0].length);
        match = pattern.exec(expr);
      }
      
      return before + String.interpret(ctx);
    });
  }
});
Template.Pattern = /(^|.|\r|\n)(#\{(.*?)\})/;


var Selector = Class.create({
  initialize: function(expression) {
    this.expression = expression.strip();
    this.compileXPathMatcher();
  },
  
  
  compileXPathMatcher: function() {
    var e = this.expression, ps = Selector.patterns,
        x = Selector.xpath, le, m, len = ps.length, name;

    if (Selector._cache[e]) {
      this.xpath = Selector._cache[e]; return;
    }

    this.matcher = ['.//*'];
    while (e && le != e && (/\S/).test(e)) {
      le = e;
      for (var i = 0; i<len; i++) {
        name = ps[i].name;
        if (m = e.match(ps[i].re)) {
          this.matcher.push(Object.isFunction(x[name]) ? x[name](m) : 
            new Template(x[name]).evaluate(m));
          e = e.replace(m[0], '');
          break;
        }
      }
    }
    
    this.xpath = this.matcher.join('')
    this.xpath = this.xpath.gsub(/\*?\[name\(\)='([a-zA-Z]+)'\]/, '#{1}');
    Selector._cache[this.expression] = this.xpath;
  },
  
  /**
   *  Selector#findElements(root) -> [Element...]
   *  - root (Element || document): A "scope" to search within. All results will
   *    be descendants of this node.
   *  
   *  Searches the document for elements that match the instance's CSS
   *  selector.
  **/
  findElements: function(root) {
    root = root || document;
    return document._getElementsByXPath(this.xpath, root);
  },
  
  /**
   *  Selector#match(element) -> Boolean
   *  
   *  Tests whether a `element` matches the instance's CSS selector.
  **/
  match: function(element) {
    this.tokens = [];

    var e = this.expression, ps = Selector.patterns, as = Selector.assertions;
    var le, p, m, len = ps.length, name;
    
    while (e && le !== e && (/\S/).test(e)) {
      le = e;
      for (var i = 0; i<len; i++) {
        p = ps[i].re;
        name = ps[i].name;
        if (m = e.match(p)) {
          // use the Selector.assertions methods unless the selector
          // is too complex.
          if (as[name]) {
            this.tokens.push([name, Object.clone(m)]);
            e = e.replace(m[0], '');
          } else {
            // reluctantly do a document-wide search
            // and look for a match in the array
            return this.findElements(document).include(element);
          }
        }
      }
    }
    
    var match = true, name, matches;
    for (var i = 0, token; token = this.tokens[i]; i++) {
      name = token[0], matches = token[1];
      if (!Selector.assertions[name](element, matches)) {
        match = false; break;
      }
    }
    
    return match;
  },
  
  toString: function() {
    return this.expression;
  },
  
  inspect: function() {
    return "#<Selector:" + this.expression.inspect() + ">";
  }
});


Object.extend(Selector, {
  _cache: { },
  
  xpath: {
    descendant:   "//*",
    child:        "/*",
    adjacent:     "/following-sibling::*[1]",
    laterSibling: '/following-sibling::*',
    tagName:      function(m) { 
      if (m[1] == '*') return '';
      //return "[local-name()='" + m[1].toLowerCase() + 
      //       "' or local-name()='" + m[1].toUpperCase() + "']";
        return "[name()='" + m[1] + "']";
    },
    className:    "[contains(concat(' ', @class, ' '), ' #{1} ')]",
    id:           "[@id='#{1}']",
    attrPresence: function(m) {
      m[1] = m[1].toLowerCase();
      return new Template("[@#{1}]").evaluate(m);
    },
    attr: function(m) {
      m[1] = m[1].toLowerCase();
      m[3] = m[5] || m[6];
      return new Template(Selector.xpath.operators[m[2]]).evaluate(m);
    },
    pseudo: function(m) {
      var h = Selector.xpath.pseudos[m[1]];
      if (!h) return '';
      if (Object.isFunction(h)) return h(m);
      return new Template(Selector.xpath.pseudos[m[1]]).evaluate(m);
    },
    operators: {
      '=':  "[@#{1}='#{3}']",
      '!=': "[@#{1}!='#{3}']",
      '^=': "[starts-with(@#{1}, '#{3}')]",
      '$=': "[substring(@#{1}, (string-length(@#{1}) - string-length('#{3}') + 1))='#{3}']",
      '*=': "[contains(@#{1}, '#{3}')]",
      '~=': "[contains(concat(' ', @#{1}, ' '), ' #{3} ')]",
      '|=': "[contains(concat('-', @#{1}, '-'), '-#{3}-')]"
    },
    pseudos: {
      'first-child': '[not(preceding-sibling::*)]',
      'last-child':  '[not(following-sibling::*)]',
      'only-child':  '[not(preceding-sibling::* or following-sibling::*)]',
      'empty':       "[count(*) = 0 and (count(text()) = 0)]",
      'checked':     "[@checked]",
      'disabled':    "[(@disabled) and (@type!='hidden')]",
      'enabled':     "[not(@disabled) and (@type!='hidden')]",
      'not': function(m) {
        var e = m[6], p = Selector.patterns,
            x = Selector.xpath, le, v, len = p.length, name;
            
        var exclusion = [];
        while (e && le != e && (/\S/).test(e)) {
          le = e;
          for (var i = 0; i<len; i++) {
            name = p[i].name
            if (m = e.match(p[i].re)) {
              v = Object.isFunction(x[name]) ? x[name](m) : new Template(x[name]).evaluate(m);
              exclusion.push("(" + v.substring(1, v.length - 1) + ")");
              e = e.replace(m[0], '');
              break;
            }
          }
        }        
        return "[not(" + exclusion.join(" and ") + ")]";
      },
      'nth-child':      function(m) { 
        return Selector.xpath.pseudos.nth("(count(./preceding-sibling::*) + 1) ", m);
      },
      'nth-last-child': function(m) {
        return Selector.xpath.pseudos.nth("(count(./following-sibling::*) + 1) ", m);
      },
      'nth-of-type':    function(m) {
        return Selector.xpath.pseudos.nth("position() ", m);
      },
      'nth-last-of-type': function(m) {
        return Selector.xpath.pseudos.nth("(last() + 1 - position()) ", m);
      },
      'first-of-type':  function(m) { 
        m[6] = "1"; return Selector.xpath.pseudos['nth-of-type'](m);
      },
      'last-of-type':   function(m) {
        m[6] = "1"; return Selector.xpath.pseudos['nth-last-of-type'](m);
      },
      'only-of-type':   function(m) {
        var p = Selector.xpath.pseudos; return p['first-of-type'](m) + p['last-of-type'](m);
      },
      nth: function(fragment, m) {
        var mm, formula = m[6], predicate;
        if (formula == 'even') formula = '2n+0';
        if (formula == 'odd')  formula = '2n+1';
        if (mm = formula.match(/^(\d+)$/)) // digit only
          return '[' + fragment + "= " + mm[1] + ']';
        if (mm = formula.match(/^(-?\d*)?n(([+-])(\d+))?/)) { // an+b
          if (mm[1] == "-") mm[1] = -1;
          var a = mm[1] ? Number(mm[1]) : 1;
          var b = mm[2] ? Number(mm[2]) : 0;
          predicate = "[((#{fragment} - #{b}) mod #{a} = 0) and " +
          "((#{fragment} - #{b}) div #{a} >= 0)]";
          return new Template(predicate).evaluate({
            fragment: fragment, a: a, b: b });
        }
      }
    }
  },

  patterns: [
    // combinators must be listed first
    // (and descendant needs to be last combinator)
    { name: 'laterSibling', re: /^\s*~\s*/ },
    { name: 'child',        re: /^\s*>\s*/ },
    { name: 'adjacent',     re: /^\s*\+\s*/ },
    { name: 'descendant',   re: /^\s/ },

    // selectors follow
    { name: 'tagName',      re: /^\s*(\*|[\w\-]+)(\b|$)?/ },
    { name: 'id',           re: /^#([\w\-\*]+)(\b|$)/ },
    { name: 'className',    re: /^\.([\w\-\*]+)(\b|$)/ },
    { name: 'pseudo',       re: /^:((first|last|nth|nth-last|only)(-child|-of-type)|empty|checked|(en|dis)abled|not)(\((.*?)\))?(\b|$|(?=\s|[:+~>]))/ },
    { name: 'attrPresence', re: /^\[((?:[\w-]+:)?[\w-]+)\]/ },
    { name: 'attr',         re: /\[((?:[\w-]*:)?[\w-]+)\s*(?:([!^$*~|]?=)\s*((['"])([^\4]*?)\4|([^'"][^\]]*?)))?\]/ }
  ],
  
  // for Selector.match and Element#match
  assertions: {
    tagName: function(element, matches) {
      return matches[1].toUpperCase() == element.tagName.toUpperCase();
    },

    className: function(element, matches) {
      return Element.hasClassName(element, matches[1]);
    },

    id: function(element, matches) {
      return element.id === matches[1];
    },

    attrPresence: function(element, matches) {
      return Element.hasAttribute(element, matches[1]);
    },

    attr: function(element, matches) {
      var nodeValue = Element.readAttribute(element, matches[1]);
      return nodeValue && Selector.operators[matches[2]](nodeValue, matches[5] || matches[6]);
    }    
  },
  operators: {
    '=':  function(nv, v) { return nv == v; },
    '!=': function(nv, v) { return nv != v; },
    '^=': function(nv, v) { return nv == v || nv && nv.startsWith(v); },
    '$=': function(nv, v) { return nv == v || nv && nv.endsWith(v); },
    '*=': function(nv, v) { return nv == v || nv && nv.include(v); },
    '~=': function(nv, v) { return (' ' + nv + ' ').include(' ' + v + ' '); },
    '|=': function(nv, v) { return ('-' + (nv || "").toUpperCase() +
     '-').include('-' + (v || "").toUpperCase() + '-'); }
  },
  
  /**
   *  Selector.split(expression) -> [String...]
   *  
   *  Takes a string of CSS selectors separated by commas; returns an array
   *  of individual selectors.
   *  
   *  Safer than doing a naive `Array#split`, since selectors can have commas
   *  in other places.
  **/
  split: function(expression) {
    var expressions = [];
    expression.scan(/(([\w#:.~>+()\s-]+|\*|\[.*?\])+)\s*(,|$)/, function(m) {
      expressions.push(m[1].strip());
    });
    return expressions;
  },

  /**
   *  Selector.matchElements(elements, expression) -> [Element...]
   *  
   *  Filters the given collection of elements with `expression`.
   *  
   *  The only nodes returned will be those that match the given CSS selector.
  **/
  matchElements: function(elements, expression) {
    var matches = $$(expression), h = Selector.handlers;
    h.mark(matches);
    for (var i = 0, results = [], element; element = elements[i]; i++)
      if (element._countedByPrototype) results.push(element);
    h.unmark(matches);
    return results;
  },
  
  /**
   *  Selector.findElement(elements, expression[, index = 0]) -> Element
   *  Selector.findElement(elements[, index = 0]) -> Element
   *  
   *  Returns the `index`th element in the collection that matches
   *  `expression`.
   *  
   *  Returns the `index`th element overall if `expression` is not given.
  **/
  findElement: function(elements, expression, index) {
    if (Object.isNumber(expression)) { 
      index = expression; expression = false;
    }
    return Selector.matchElements(elements, expression || '*')[index || 0];
  },
  
  /**
   *  Selector.findChildElements(element, expressions) -> [Element...]
   *  
   *  Searches beneath `element` for any elements that match the selector
   *  (or selectors) specified in `expressions`.
  **/
  findChildElements: function(element, expressions) {
    expressions = Selector.split(expressions.join(','));
    var results = [], h = Selector.handlers;    
    for (var i = 0, l = expressions.length, selector; i < l; i++) {
      selector = new Selector(expressions[i].strip());
      h.concat(results, selector.findElements(element));
    }
    return (l > 1) ? h.unique(results) : results;
  }
});

File: ayril/selector.rb

class Selector
  attr_reader :expr, :xpath
  @@cache = {}

  def initialize(expr)
    @expr = expr.strip
    self.compile_xpath_matcher
  end
  
  def compile_xpath_matcher
    e = @expr.dup; le = nil
    return (@xpath = @@cache[e]) if @@cache.include? e

    @matcher = [".//*"]
    while (e != '') and (le != e) and (e =~ /\S/)
      le = e.dup
      Selector::Patterns.each do |pattern|
        n = Selector::XPath[pattern[:name]]
        if m = e.match(pattern[:re])
          m = m.to_a unless m.nil?
          @matcher << (n.kind_of?(Proc) ? n.call(m) : n.interpolate(m))
          e.sub! m[0], ''
          break
        end
      end
    end

    @xpath = @matcher.join
    @@cache[@expr] = @xpath.gsub! %r{\*?\[name\(\)='([a-zA-Z]+)'\]}, '\1'
  end

  def find_elements(root)
    root.select_by_xpath @xpath
  end

  def match?(element)
    @tokens = []
    
    e = @expr.dup; le = nil
    while (e != '') and (le != e) and (e =~ /\S/)
      le = e.dup
      Selector::Patterns.each do |pattern|
        if m = e.patch(pattern[:re])
          m = m.to_a unless m.nil?
          if Selector::Assertions.include? name
            @tokens << [pattern[:name], m.clone]
            e.sub! m[0], ''
          else # resort to whole document
            return self.find_elements(element.rootDocument).include? element
          end
        end
      end
    end

    match = true
    @tokens.each do |token|
      name, matches = token[0..1]
      if not Selector::Assertions[name].call self, matches
        match = false
        break
      end
    end
    match
  end

  def to_s
    @expr
  end

  def inspect
    "#<Selector:#{@expr.inspect}>"
  end
  

  XPath = {
    :descendant =>   "//*",
    :child =>        "/*",
    :adjacent =>     "/following-sibling::*[1]",
    :laterSibling => '/following-sibling::*',
    :tagName =>      lambda { |m|
      return '' if m[1] == '*'
      "[name()='" + m[1] + "']"
    },
    :className =>    "[contains(concat(' ', @class, ' '), ' \#{1} ')]",
    :id =>           "[@id='\#{1}']",
    :attrPresence => lambda { |m|
      m[1].downcase!
      "[@\#{1}]".interpolate(m)
    },
    :attr => lambda { |m|
      m[1].downcase!
      m[3] = m[5] || m[6]
      Selector::XPath[:operators][m[2]].interpolate(m)
    },
    :pseudo => lambda { |m|
      h = Selector::XPath[:pseudos][m[1]]
      return '' if h.nil?
      return h.call(m) if h.kind_of? Proc
      Selector::XPath[:pseudos][m[1]].interpolate(m)
    },
    :operators => {
      '='  => "[@\#{1}='\#{3}']",
      '!=' => "[@\#{1}!='\#{3}']",
      '^=' => "[starts-with(@\#{1}, '\#{3}')]",
      '$=' => "[substring(@\#{1}, (string-length(@\#{1}) - string-length('\#{3}') + 1))='\#{3}']",
      '*=' => "[contains(@\#{1}, '\#{3}')]",
      '~=' => "[contains(concat(' ', @\#{1}, ' '), ' \#{3} ')]",
      '|=' => "[contains(concat('-', @\#{1}, '-'), '-\#{3}-')]"
    },
    :pseudos => {
      'first-child' => '[not(preceding-sibling::*)]',
      'last-child' =>  '[not(following-sibling::*)]',
      'only-child' =>  '[not(preceding-sibling::* or following-sibling::*)]',
      'empty' =>       "[count(*) = 0 and (count(text()) = 0)]",
      'checked' =>     "[@checked]",
      'disabled' =>    "[(@disabled) and (@type!='hidden')]",
      'enabled' =>     "[not(@disabled) and (@type!='hidden')]",
      'not' => lambda { |m|
        e = m[6]; le = nil; exclusion = []
        while (e != '') and (le != e) and (e =~ /\S/)
          le = e.dup
          Selector::Patterns.each do |pattern|
            n = Selector::XPath[pattern[:name]]
            if m = e.match(pattern[:re])
              v = n.kind_of?(Proc) ? n.call(m) : n.interpolate(m)
              exclusion << ('(' + v[1, v.length - 2] + ')')
              e.gsub! m[0], ''
              break
            end
          end
        end      
        "[not(" + exclusion.join(" and ") + ")]"
      },
      'nth-child' =>      lambda { |m| 
        Selector::XPath[:pseudos]['nth'].call("(count(./preceding-sibling::*) + 1) ", m)
      },
      'nth-last-child' => lambda { |m|
        Selector::XPath[:pseudos]['nth'].call("(count(./following-sibling::*) + 1) ", m)
      },
      'nth-of-type' =>    lambda { |m|
        Selector::XPath[:pseudos]['nth'].call("position() ", m)
      },
      'nth-last-of-type' => lambda { |m|
        Selector::XPath[:pseudos]['nth'].call("(last() + 1 - position()) ", m)
      },
      'first-of-type' =>  lambda { |m| 
        m[6] = "1"; Selector::XPath[:pseudos]['nth-of-type'].call(m)
      },
      'last-of-type' =>   lambda { |m|
        m[6] = "1"; Selector::XPath[:pseudos]['nth-last-of-type'].call(m)
      },
      'only-of-type' =>   lambda { |m|
        p = Selector::XPath[:pseudos]
        p['first-of-type'].call(m) + p['last-of-type'].call(m)
      },
      'nth' => lambda { |fragment, m|
        mm = nil; formula = m[6]; predicate = nil
        formula = '2n+0' if formula == 'even'
        formula = '2n+1' if formula == 'odd'
        return "[#{fragment}= #{mm[1]}]" if mm = formula.match(/^(\d+)$/) # digit only
        if mm = formula.match(/^(-?\d*)?n(([+-])(\d+))?/).to_a # an+b
          mm[1] = -1 if mm[1] == '-'
          # in JS: mm => ['n', undefined, undefined, undefined, undefined]
          # in RB: mm => ['n', '', nil, nil, nil]
          mm[1] = nil if mm[0] == 'n' # edge case
          a = mm[1] ? mm[1].to_i : 1
          b = mm[2] ? mm[2].to_i : 0
          "[((#{fragment} - #{b}) mod #{a} = 0) and ((#{fragment} - #{b}) div #{a} >= 0)]"
        end
      }
    }    
  }

  Patterns = [
    # combinators must be listed first
    # (and descendant needs to be last combinator)
    { :name => :laterSibling, :re => %r{^\s*~\s*}  },
    { :name => :child,        :re => %r{^\s*>\s*}  },
    { :name => :adjacent,     :re => %r{^\s*\+\s*}  },
    { :name => :descendant,   :re => %r{^\s}  },

    # selectors follow
    { :name => :tagName,      :re => %r{^\s*(\*|[\w\-]+)(\b|$)?}  },
    { :name => :id,           :re => %r{^#([\w\-\*]+)(\b|$)}  },
    { :name => :className,    :re => %r{^\.([\w\-\*]+)(\b|$)}  },
    { :name => :pseudo,       :re => %r{^:((first|last|nth|nth-last|only)(-child|-of-type)|empty|checked|(en|dis)abled|not)(\((.*?)\))?(\b|$|(?=\s|[:+~>]))}  },
    { :name => :attrPresence, :re => %r{^\[((?:[\w-]+:)?[\w-]+)\]}  },
    { :name => :attr,         :re => %r{\[((?:[\w-]*:)?[\w-]+)\s*(?:([!^$*~|]?=)\s*((['"])([^\4]*?)\4|([^'"][^\]]*?)))?\]} }
  ]

  Assertions = {
    "tagName" => lambda { |element, matches|
       matches[1].upcase == element.name.upcase
    },

    "className" => lambda { |element, matches|
       element.has_class_name? matches[1]
    },

    "id" => lambda { |element, matches|
       element.read_attribute("id") == matches[1]
    },

    "attrPresence" => lambda { |element, matches|
       element.has_attribute? matches[1]
    },

    "attr" => lambda { |element, matches|
      value = element.read_attribute matches[1]
      value and Selector::Operators[matches[2]].call(value, matches[5] || matches[6])
    }
  }

  Operators = {
    '=' =>  lambda { |nv, v| nv == v },
    '!=' => lambda { |nv, v| nv != v },
    '^=' => lambda { |nv, v| nv == v or (nv and nv.start_with? v) },
    '$=' => lambda { |nv, v| nv == v or (nv and nv.end_with? v) },
    '*=' => lambda { |nv, v| nv == v or (nv and nv.include? v) },
    '~=' => lambda { |nv, v| " #{nv} ".include? " #{v} " },
    '|=' => lambda { |nv, v| "-#{(nv || '').upcase}-".include? "-#{(v || '').upcase}-" }    
  }

  def self.split(expression)
    expressions = []
    expression.scan(/(([\w#:.~>+()\s-]+|\*|\[.*?\])+)\s*(,|$)/) do |m|
      expressions << m[1].strip
    end
    expressions
  end

  def self.match_elements(elements, expression)
    elements & elements[0].rootDocument.select(expression)
  end

  def self.find_element(elements, *rest)
    expression, index = rest[0], rest[1]
    (expression = nil; index = expression) if expression.kind_of? Integer
    Selector::match_elements(elements, expression || '*')[index || 0]
  end

  def self.find_child_elements(element, expressions)
    Selector::split(expressions.join(',')).map do |expression|
      Selector.new(expression.strip).find_elements(element)
    end.uniq.flatten
  end
end



